### ISSUE TITLE
Related keyof T types are not assignable to each other

### ISSUE THREAD

**Issue by MaartenStaa** [microsoft/TypeScript#30828]
**TypeScript Version:**  3.4.2

**Search Terms:** Generics, keyof, strict function types

**Code**

```ts
interface Foo {
  [type: string]: any // tslint:disable-line no-any
}

type Bar<T, K extends keyof T> = (_: T[K]) => void

class Baz<T extends Foo, K extends keyof T = keyof T> {
  private list: { [key: string]: Bar<T, K>[] } = {}

  callMe<E extends K>(key: E, item: Bar<T, E>): void {
    // Workaround: Type 'E' cannot be used to index type '{ [key: string]: Bar<T, K>[]; }'
    const stringKey: string = key as string
    if (! (key in this.list)) {
      this.list[stringKey] = []
    }

    this.list[stringKey].push(item)
  }
}
```

**Expected behavior:** No type errors.

**Actual behavior:** An error on line 17:

```
Argument of type 'Bar<T, E>' is not assignable to parameter of type 'Bar<T, K>'.
  Type 'K' is not assignable to type 'E'.
    Type 'keyof T' is not assignable to type 'E'.
      Type 'string | number | symbol' is not assignable to type 'E'.
        Type 'string' is not assignable to type 'E'.
```

This same code worked fine on Typescript 3.3, but started emitting errors after upgrading to 3.4. The error only occurs when strict function types is enabled.

**Playground Link:** https://www.typescriptlang.org/play/index.html#src=interface%20Foo%20%7B%0D%0A%20%20%5Btype%3A%20string%5D%3A%20any%20%2F%2F%20tslint%3Adisable-line%20no-any%0D%0A%7D%0D%0A%0D%0Atype%20Bar%3CT%2C%20K%20extends%20keyof%20T%3E%20%3D%20(_%3A%20T%5BK%5D)%20%3D%3E%20void%0D%0A%0D%0Aclass%20Baz%3CT%20extends%20Foo%2C%20K%20extends%20keyof%20T%20%3D%20keyof%20T%3E%20%7B%0D%0A%20%20private%20list%3A%20%7B%20%5Bkey%3A%20string%5D%3A%20Bar%3CT%2C%20K%3E%5B%5D%20%7D%20%3D%20%7B%7D%0D%0A%0D%0A%20%20callMe%3CE%20extends%20K%3E(key%3A%20E%2C%20item%3A%20Bar%3CT%2C%20E%3E)%3A%20void%20%7B%0D%0A%20%20%20%20%2F%2F%20Workaround%3A%20Type%20'E'%20cannot%20be%20used%20to%20index%20type%20'%7B%20%5Bkey%3A%20string%5D%3A%20Bar%3CT%2C%20K%3E%5B%5D%3B%20%7D'%0D%0A%20%20%20%20const%20stringKey%3A%20string%20%3D%20key%20as%20string%0D%0A%20%20%20%20if%20(!%20(key%20in%20this.list))%20%7B%0D%0A%20%20%20%20%20%20this.list%5BstringKey%5D%20%3D%20%5B%5D%0D%0A%20%20%20%20%7D%0D%0A%0D%0A%20%20%20%20this.list%5BstringKey%5D.push(item)%0D%0A%20%20%7D%0D%0A%7D%0D%0A

**Related Issues:** 

#21756 seems similar, but not the same exactly.

**Comment by jack-williams** [comment]
I think the error is correct here, otherwise you can end up with unsoundness. E.g.
```ts
interface Foo {
  [type: string]: any // tslint:disable-line no-any
}

type Bar<T, K extends keyof T> = (_: T[K]) => void

class Baz<T extends Foo, K extends keyof T = keyof T> {
  private list: { [key: string]: Bar<T, K>[] } = {}

  callMe<E extends K>(key: E, item: Bar<T, E>): void {
    // Workaround: Type 'E' cannot be used to index type '{ [key: string]: Bar<T, K>[]; }'
    const stringKey: string = key as string
    if (! (key in this.list)) {
      this.list[stringKey] = []
    }

    this.list[stringKey].push(item)
  }

  applyMe(key: string) {
    return this.list[key][0];
  }

}

declare const baz: Baz<{ x: number; y: string; z: boolean }, "y" | "z">
baz.callMe("y", (x: string) => undefined);
baz.applyMe("y")(true); // the function was set as something that only accepted strings!
```
The issue appears to be that you're assigning a `Bar<T, E>` to a `Bar<T, K>` when `E` is related to `K`, however both those types are used as keys, which are related contravariantly. That is, for `Bar<T, E>` to be related to `Bar<T, K>`, is should be that `K` is related to `E`.
